File: backend/crates/domain/coffret-usecase/src/sync_conformance/unmapped.rs
use crate::sync::sync_folders;
use crate::sync_conformance::fixtures::{keys, map, request, write};
use crate::sync_conformance::sync_under_test::SyncUnderTest;

/// A device that maps nothing is told apart from folders with nothing new.
///
/// The two produce the same counts — nothing added, nothing replaced, nothing
/// committed — and they are different states. A run that found every file
/// already current is the ordinary second sync of a folder. A device with no
/// mapping at all has nothing inside the Library's scope, so it would answer
/// that way whatever is on its disk, and somebody who has just joined a Library
/// would read it as everything already being backed up (spec: EP-9).
///
/// So the run says how many mappings it worked from, and a caller that shows
/// the counts alone has the one fact that tells the two apart. It is the same
/// fact a fetch reports for the same reason, because the person reading it is
/// the same person: whoever has just joined tries one of the two commands
/// first, and which one they try must not decide whether they are told.
pub async fn a_device_that_maps_nothing_is_told_apart_from_a_folder_with_nothing_new(
    fixture: &SyncUnderTest,
) {
    let keys = keys();
    let store = fixture.store();
    let index = fixture.index();

    // Files on disk and no mapping at all, which is what a device that has only
    // just joined looks like.
    write(fixture.fs(), fixture.folder(), "a.jpg", b"a photo");

    let unmapped = sync_folders(request(store, index, &keys, fixture.fs(), 1))
        .await
        .unwrap_or_else(|error| {
            panic!("a sync by a device that maps nothing must succeed: {error}")
        });
    assert_eq!(unmapped.mappings, 0, "the device has recorded no mapping");
    assert!(
        unmapped.added.is_empty(),
        "nothing on this disk is inside the Library's scope",
    );
    assert!(
        unmapped.commit.is_none(),
        "a run with nothing to upload commits nothing (spec: CP-1)",
    );

    map(fixture, None).await;
    let carried = sync_folders(request(store, index, &keys, fixture.fs(), 2))
        .await
        .unwrap_or_else(|error| panic!("a sync of a mapped folder must succeed: {error}"));
    assert_eq!(carried.mappings, 1, "the device maps a folder now");
    assert_eq!(carried.added.len(), 1, "and the file it holds went up");

    // The ordinary second run: the same empty counts as the first run above,
    // and the one number that says why they are empty differs.
    let again = sync_folders(request(store, index, &keys, fixture.fs(), 3))
        .await
        .unwrap_or_else(|error| panic!("a second sync of the folder must succeed: {error}"));
    assert_eq!(
        again.mappings, 1,
        "the device maps a folder; the folder is what held nothing new",
    );
    assert!(again.added.is_empty());
    assert!(again.commit.is_none());
}

File: frontend/packages/apps/web/src/unmapped.ts
// What a gesture is answered with in a folder this device can do nothing in.
//
// The banner over such a folder is the standing reason it is there at all.
// These are the answers to the things that were just tried in front of it, and
// a gesture that changes nothing on the screen needs one: a row clicked and a
// drop let go of both look, a second later, exactly like a screen that was
// never touched.
//
// Two folders reach here and not one: a folder of the Library no mapping of
// this device reaches, and a path the Library holds nothing at. The gestures
// fail in the same place — nothing to fetch, nowhere to put it — so the answer
// is written here for both, and which reason it gives is the whole of what
// `held` decides.
//
// Kept free of DOM and of React so it is unit testable, the way the drop's own
// sentences beside it are.

/** Which gesture met the folder. */
export type Tried = 'open' | 'add';

/**
 * The line the notice area shows, naming what was tried and why it came to
 * nothing.
 *
 * Two sentences rather than one, differing in the half that says what was
 * tried. The reason is the same fact about the same folder, and it is worded the
 * same both times, so that reading the second after the first is reading one
 * new clause rather than a new sentence. What differs is everything a person
 * can check: they clicked a row, or they let go of files, and a line beginning
 * "nothing was added" under a row they clicked would have them looking for the
 * drop they did not make.
 *
 * Which fact it is, is `held`'s to say, and one path never gets both. A folder
 * the Library holds that no mapping of this device reaches is the EP-9 reason:
 * the folder is there, and this device has nowhere to put what is in it. A path
 * the Library holds nothing at is the other, and mapping is not what is wrong
 * with it — the screen over these rows deliberately says nothing about mapping
 * such a path, since being told to map a part of the Library there is none of
 * sends somebody to a terminal for nothing, and a notice that gave the mapping
 * reason anyway would leave one path carrying two explanations at once.
 *
 * `below` is the folders inside this one that this device *does* have a folder
 * for, which is only ever a Library root's children: a mapping stands for one
 * top-level component and everything under it, so deeper down every child
 * repeats its parent's answer. Naming them is the whole of what this adds to
 * the banner. A drop of a nested folder onto an unmapped root is refused entire
 * although a mapped folder below the root could have taken its share, and a
 * refusal that does not say which part was refusable leaves somebody to work
 * out by trial which of their folders the device has. It goes with the EP-9
 * reason alone, and not by being held back: a folder of the Library is what the
 * paths under it imply, so a path the Library does not hold has no children of
 * any kind, mapped or otherwise.
 *
 * Refused entire, and not split. Taking the part that has somewhere to go is
 * the better outcome for a person and is a different gesture underneath: the
 * drop is one request against one folder, so a half-accepted drop is several
 * requests, a rule for which files each carries, and an answer that has to
 * report what landed and what did not from more than one place. What is wrong
 * here is that the refusal says nothing about the folders, and that is wrong
 * under either shape — so it is fixed first, and on its own.
 *
 * It says nothing about `below` for an `open`. The row that was clicked stands
 * in *this* folder, and no folder beside it can hold that file: there is no
 * other place to offer, and offering one would be inviting somebody to look for
 * their file where it is not.
 */
export function unmappedLine(
  tried: Tried,
  held: boolean,
  below: readonly string[],
): string {
  if (!held) {
    // One clause for both gestures, because there is nothing further to say
    // about either: no mapping would change this, and there is no folder below
    // to offer. What was tried is the only half that differs.
    const missing = 'the Library holds nothing at this path';
    return tried === 'open'
      ? `nothing was opened — ${missing}`
      : `nothing was added — ${missing}`;
  }
  const reason = 'no folder on this device holds this part of the Library';
  if (tried === 'open') {
    return `nothing was opened — ${reason}, so there is nowhere here to put the file that row stands for`;
  }
  if (below.length === 0) {
    return `nothing was added — ${reason}`;
  }
  return `nothing was added — ${reason}; the folders below it that do are ${named(below)}, and a drop onto one of those is taken`;
}

/**
 * The folders, as a list a sentence can carry.
 *
 * Two of them and a count of the rest, which is the economy the drop's own
 * refusals keep: the notice area is one line, and a Library root of thirty
 * top-level folders would otherwise put a paragraph in it. Two rather than one
 * because the point of naming any is that a person recognizes theirs, and one
 * name reads as the only answer.
 */
function named(below: readonly string[]): string {
  if (below.length === 1) {
    return below[0];
  }
  if (below.length === 2) {
    return `${below[0]} and ${below[1]}`;
  }
  return `${below[0]}, ${below[1]} and ${below.length - 2} more`;
}
